membership-attribute-service/app/services/AttributesFromZuora.scala

```diff
@@ -50,24 +52,30 @@ class AttributesFromZuora(implicit val executionContext: ExecutionContext) exten
           case Failure(e) => SafeLogger.error(scrub"Failed to retrieve attributes from cache for $identityId because $e")
         }
 
+    lazy val getAttrFromZuora =
+      for {
+        accounts <- EitherT(identityIdToAccounts(identityId))
+        subscriptions <- EitherT(if (accounts.size > 0) getSubscriptions(accounts, subscriptionsForAccountId) else Future.successful(\/.right[String, List[AccountWithSubscriptions]](Nil)))
+        maybeZAttributes <- EitherT(AttributesMaker.zuoraAttributes(identityId, subscriptions, paymentMethodForPaymentMethodId, forDate).map(\/.right[String, Option[ZuoraAttributes]]))
+      } yield {
+        val maybeAttributes = maybeZAttributes.map(ZuoraAttributes.asAttributes(_))
+        attributesFromDynamo.foreach(maybeUpdateCache(maybeZAttributes, _, maybeAttributes))
+        Future.successful("Zuora", maybeAttributes)
+      }
+
     lazy val fallbackToCache = (zuoraError: String) => {
       SafeLogger.error(scrub"Failed to retrieve attributes from Zuora for $identityId because $zuoraError so falling back on cache.")
       attributesFromDynamo.map(maybeDynamoAttributes => ("Dynamo", maybeDynamoAttributes.map(DynamoAttributes.asAttributes(_))))
     }
 
-    (for { // try zuora first and fallback to cache
-      accounts         <- EitherT(identityIdToAccounts(identityId))
-      subscriptions    <- EitherT(if (accounts.size > 0) getSubscriptions(accounts, subscriptionsForAccountId) else Future.successful(\/.right[String, List[AccountWithSubscriptions]](Nil)))
-      maybeZAttributes <- EitherT(AttributesMaker.zuoraAttributes(identityId, subscriptions, paymentMethodForPaymentMethodId, forDate).map(\/.right[String, Option[ZuoraAttributes]]))
-    } yield { // Zuora takes precedence
-      val maybeAttributes = maybeZAttributes.map(ZuoraAttributes.asAttributes(_))
-      attributesFromDynamo.foreach(maybeUpdateCache(maybeZAttributes, _, maybeAttributes))
-      Future.successful("Zuora", maybeAttributes)
-    }) // fallback to cache if Zuora networking error or un-parsable response
-      .leftMap(fallbackToCache) // handles un-parsable response etc.
-      .merge // both branches return the same type so we can merge
+    implicit val scheduler: Scheduler = system.scheduler
+
+    // try zuora first and fallback to cache
+    EitherT(retry(getAttrFromZuora.run))                     // Zuora takes precedence
```

Comment:
Is one retry enough?

Reply:
Increased it to 2. Not sure if there is much point beyond that.